include/xraylib-error.h

```diff
@@ -0,0 +1,83 @@
+/* Copyright (C) 2018 Tom Schoonjans
+All rights reserved.
+
+Redistribution and use in source and binary forms, with or without
+modification, are permitted provided that the following conditions are met:
+    * Redistributions of source code must retain the above copyright notice, this list of conditions and the following disclaimer.
+    * Redistributions in binary form must reproduce the above copyright notice, this list of conditions and the following disclaimer in the documentation and/or other materials provided with the distribution.
+    * The names of the contributors may not be used to endorse or promote products derived from this software without specific prior written permission.
+
+THIS SOFTWARE IS PROVIDED BY Tom Schoonjans 'AS IS' AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL Tom Schoonjans BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+*/
+
+#ifndef XRAYLIB_ERROR_H
+#define XRAYLIB_ERROR_H
+
+#include <stdarg.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/*
+ *  This file is mostly copy-pasted from GLib's error methods...
+ */ 
+
+#if __GNUC__ > 2 || (__GNUC__ == 2 && __GNUC_MINOR__ > 4)
+#define GNUC_PRINTF( format_idx, arg_idx )    \
+  __attribute__((__format__ (__printf__, format_idx, arg_idx)))
+#else /* !__GNUC__ */
+#define GNUC_PRINTF( format_idx, arg_idx )
+#endif /* !__GNUC__ */
+
+enum xrl_error_code {
+	XRL_ERROR_MEMORY, /* set in case of a memory allocation problem */
+	XRL_ERROR_INVALID_ARGUMENT, /* set in case an invalid argument gets passed to a routine */
+	XRL_ERROR_IO, /* set in case an error involving input/output occurred */
+	XRL_ERROR_TYPE, /* set in case an error involving type conversion occurred */
+	XRL_ERROR_UNSUPPORTED, /* set in case an unsupported feature has been requested */
+	XRL_ERROR_RUNTIME, /* set in case an unexpected runtime error occurred */
+};
+
+
+/**
+ * xrl_error:
+ * @code: error code, e.g. %XRL_ERROR_MEMORY
+ * @message: human-readable informative error message
+ *
+ * The `xrl_error` structure contains information about
+ * an error that has occurred.
+ */
+typedef struct _xrl_error xrl_error;
+
+struct _xrl_error
+{
+  enum xrl_error_code code;
+  char *message;
+};
+
+xrl_error* xrl_error_new(enum xrl_error_code code, const char *format, ...) GNUC_PRINTF (2, 3);
```

Comment:
The code looks good, but I believe the error construction routines should not be exported from the library. Only the library should create errors and the user of the library should only need xrl_error_free and xrl_clear_error.

Reply:
Yep will move those to a private header and make sure they're not exported from the library (see #83 )